Libplanet/Blockchain/BlockChain.cs

```diff
@@ -1792,6 +1781,69 @@ internal InvalidBlockException ValidateNextBlock(Block<T> block)
             return null;
         }
 
+        /// <summary>
+        /// Validates the contents of the given <see cref="Block{T}"/>.
+        /// </summary>
+        /// <param name="block">A block to be validated.</param>
+        /// <param name="evaluateActions">Pre-evaluated actions that corresponding to the given
+        /// block. If <see langword="null"/> is given, it will evaluate the block action.</param>
+        /// <param name="writeTxExecution">Whether writing the transaction execution results
+        /// <see cref="TxExecution"/>, that paired with block, to the store. Set this
+        /// <see langword="false"/> if any execution result does not need to be written
+        /// (e.g., rehearsal, validation).
+        /// </param>
+        /// <returns>The result of action evaluations and the difference of Tx nonces of the given
+        /// <paramref name="block"/>.</returns>
+        /// <exception cref="TxPolicyViolationException">Thrown if any given transaction in the
+        /// block violates the defined <see cref="Policy"/>.
+        /// </exception>
+        /// <exception cref="InvalidTxNonceException">Thrown if any given transaction in the
+        /// block has an invalid nonce.</exception>
+        /// <seealso cref="ValidateNonces"/>
+        /// <seealso cref="Policies.IBlockPolicy{T}.ValidateNextBlockTx"/>
+        /// <seealso cref="ExecuteActions"/>
+        internal (IReadOnlyList<ActionEvaluation> ActionEvaluations,
+            Dictionary<Address, long> NonceDelta) ValidateNextBlockContent(
+                Block<T> block,
+                bool evaluateActions,
+                bool writeTxExecution = false)
+        {
+            var nonceDeltas = ValidateNonces(
+                    block.Transactions
+                        .Select(tx => tx.Signer)
+                        .Distinct()
+                        .ToDictionary(signer => signer, signer => Store.GetTxNonce(Id, signer)),
+                    block);
+
+            foreach (Transaction<T> tx in block.Transactions)
+            {
+                if (Policy.ValidateNextBlockTx(this, tx) is { } tpve)
+                {
+                    throw new TxPolicyViolationException(
+                        "According to BlockPolicy, this transaction is not valid.",
+                        tx.Id,
+                        tpve);
+                }
+            }
+
+            if (evaluateActions)
+            {
+                _logger.Information(
+                    "Executing actions in block #{BlockIndex} {BlockHash}...",
+                    block.Index,
+                    block.Hash);
+                var actionEvaluations = ExecuteActions(block, writeTxExecution);
+                _logger.Information(
+                    "Executed actions in block #{BlockIndex} {BlockHash}",
+                    block.Index,
+                    block.Hash);
+
+                return (actionEvaluations, nonceDeltas);
+            }
```

Comment:
I have general qualms with evaluating the content inside `ValidateNextBlockContent()`. 🙄

- Evaluation result isn't part of `BlockContent<T>`'s scope.
- I think usage `writeTxExecution` for the sake of having an ephemeral execution is bad. Actually `UpdateTxExecutions()` isn't where it should be (that is, inside `EvaluateActions()` instead of `Append()`).
- Combining the two, the signature for this should be
```csharp
Dictionary<Address, long> ValidateNextBlockContent(IBlockContent<T>)
```